### PR TITLE
Preparation for release 0.1.2

### PR DESCRIPTION
**Pull request by Dawid64**
After this pull request will be merged I will create new release (Version 0.1.2)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,21 +14,22 @@ Install Decorators via pip:
...
+| timeit | Measures the time of the function |
+| timeout | Terminates the function after it runs for too long |
+| mute | Disables stdout for duration of decorated function |
+| validate_typehints | Ensures that typehints are followed, raising error if not |
+| retry | If exception is raised function is retried up to predefined amount of times |
+| rate_limiter | Limits number of function calls such that there are no function calls within set amount of seconds |
+| time_limiter | Splits time into intervals, and limits the number of function calls within each interval |
+| default_value | Set default value for function if exception is raised |
+| grid_search | Allow user to perform grid search on each iteration of function call |
+| redirect | Enables user to redirect function input / output |
```
Comment (adamtms) [inline README.md:29]: It should be something like redirect functions stdout/stderr. This doesnt redirect input.

**Review by adamtms — CHANGES_REQUESTED** [review]
I am also wondering if time limiter is a correct name. It is kind of misleading, because from the name I would assume it works similarily to timeout. 
E.g. it could be tumbling_rate_limiter since it uses tumbling window, however I wouldnt be surprised if someone else had a better idea.
Other than this and the other comment it looks good.

File: README.md (hunk 1 above)
Comment (Dawid64) [reply README.md:29]: Yep, your right, my bad

**Comment by Dawid64** [comment]
interval_rate_limitter maybe isn't as correct as tumbling, but I think it's easier to find by name, I would first think about interval, if looking for such decorator, if not, you can propose other name

**Review by JaszczurGra — CHANGES_REQUESTED** [review]
Both tests for redirecting stdout/stderr return an error on windows:
```py
ValueError: Invalid file type. Please pass a file-like object, a string or None.
```

**Comment by JaszczurGra** [comment]
Why are not all of the functions in the readme table (like loop, average)?

There is a  **TODO** in profiling.crawler, idk if it should be resolved before the merge

> **TODO:** Add support for exceptions

In description of iterative_time_limiter the sentence doesn't make sense (not)

> **!!Note that this decorator is not thread-safe and should **not** be used carefully in a multi-threaded environment!!**

As well as this sentence 
> The decorator can also synchronize with the system  clock and reset the call counter each time interval counting from 00:00:00.

After reading the code it makes more sense but still sounds confusing for the user, doesn't it break when the program runs for more than 24h? 



All the other tests pass on windows and I really like the new table in the readme

**Comment by Dawid64** [comment]
> Why are not all of the functions in the readme table (like loop, average)?

I was thinking to upgrade them before releasing, adding such features as  multiprocessing, to make them even more useful, but I think it's better to make version 0.1.2 sooner than later, for easier using the library

> There is a **TODO** in profiling.crawler, idk if it should be resolved before the merge
> > **TODO:** Add support for exceptions

TODO does not need to be resolved yet

> In description of iterative_time_limiter the sentence doesn't make sense (not)
> > **!!Note that this decorator is not thread-safe and should not be used carefully in a multi-threaded environment!!**
> As well as this sentence
> > The decorator can also synchronize with the system  clock and reset the call counter each time interval counting from 00:00:00.

Fixed

> After reading the code it makes more sense but still sounds confusing for the user, doesn't it break when the program runs for more than 24h?

It will work, it just set starting value to 00:00:00 of current day, but I believe it should work for more than 1 day

> Both tests for redirecting stdout/stderr return an error on windows:
> 
> ```python
> ValueError: Invalid file type. Please pass a file-like object, a string or None.
> ```

There are 6 tests for redirecting stdout/stderr, and they worked on my machine / on GitHub workflow, can you double check if you're in correct branch and if so, what is the issue and why does it happen? If you would not be able to solve the problem you can show whole error message here for failed tests.

**Comment by JaszczurGra** [comment]
> I was thinking to upgrade them before releasing, adding such features as multiprocessing, to make them even more useful, but I think it's better to make version 0.1.2 sooner than later, for easier using the library

They could still be shown in the readme even if they are not in the final form, they are still useful




> There are 6 tests for redirecting stdout/stderr, and they worked on my machine / on GitHub workflow, can you double check if you're in correct branch and if so, what is the issue and why does it happen? If you would not be able to solve the problem you can show whole error message here for failed tests.

The 2 I that are not working are **test_redirect_stdout_file** and **test_redirect_stderr_file**
The error for test_redirect_stdout_file is:
```c:\Programing\Repos\Decorators\tests\test_io_redirect.py::test_redirect_stdout_file failed: capsys = <_pytest.capture.CaptureFixture object at 0x000001C0674A9F90>

    def test_redirect_stdout_file(capsys):
        with tempfile.TemporaryFile("w+") as f:
            @redirect(stdout_target=f)
            def print_hello():
                print("Hello, world!")
    
>           print_hello()

test_io_redirect.py:49: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\decorify\io_redirect.py:87: in inner_func
    dest_stdout = __redirect_dest(stdout_target)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

file = <tempfile._TemporaryFileWrapper object at 0x000001C067496750>

    def __redirect_dest(file: Union[IOBase, str, None]) -> IOBase:
        """ Returns a file-like object to redirect stdout to."""
        if file is None:
            return open(os.devnull, 'w')
        elif isinstance(file, str):
            return open(file, 'a')
        elif isinstance(file, IOBase):
            return file
>       raise ValueError("Invalid file type. Please pass a file-like object, a string or None.")
E       ValueError: Invalid file type. Please pass a file-like object, a string or None.

..\decorify\io_redirect.py:55: ValueError
```
So the problem is with the tempfile

**Comment by adamtms** [comment]
> > I was thinking to upgrade them before releasing, adding such features as multiprocessing, to make them even more useful, but I think it's better to make version 0.1.2 sooner than later, for easier using the library
> 
> They could still be shown in the readme even if they are not in the final form, they are still useful
> 
> > There are 6 tests for redirecting stdout/stderr, and they worked on my machine / on GitHub workflow, can you double check if you're in correct branch and if so, what is the issue and why does it happen? If you would not be able to solve the problem you can show whole error message here for failed tests.
> 
> The 2 I that are not working are **test_redirect_stdout_file** and **test_redirect_stderr_file** The error for test_redirect_stdout_file is:
> 
> ```
> 
>     def test_redirect_stdout_file(capsys):
>         with tempfile.TemporaryFile("w+") as f:
>             @redirect(stdout_target=f)
>             def print_hello():
>                 print("Hello, world!")
>     
> >           print_hello()
> 
> test_io_redirect.py:49: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> ..\decorify\io_redirect.py:87: in inner_func
>     dest_stdout = __redirect_dest(stdout_target)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> file = <tempfile._TemporaryFileWrapper object at 0x000001C067496750>
> 
>     def __redirect_dest(file: Union[IOBase, str, None]) -> IOBase:
>         """ Returns a file-like object to redirect stdout to."""
>         if file is None:
>             return open(os.devnull, 'w')
>         elif isinstance(file, str):
>             return open(file, 'a')
>         elif isinstance(file, IOBase):
>             return file
> >       raise ValueError("Invalid file type. Please pass a file-like object, a string or None.")
> E       ValueError: Invalid file type. Please pass a file-like object, a string or None.
> 
> ..\decorify\io_redirect.py:55: ValueError
> ```
> 
> So the problem is with the tempfile

I fixed the issue. For some weird reason tempfile on windows inherits directly from object. I changed the implementation to validate whether there exist both read and write method on the passed variable instead of checking inheritance when user passes a file.

**Review by adamtms — APPROVED** [review]

**Review by JaszczurGra — APPROVED** [review]
**Lgtm** <3

**Review by adamtms — APPROVED** [review]
LGTM

**Pull request merged**